cabal-dev-scripts/src/GenValidate.hs

```diff
@@ -1,159 +0,0 @@
-module Main (main) where
-
-import Data.List          (isPrefixOf)
-import Data.Maybe         (fromMaybe)
-import GHC.Generics       (Generic)
-import System.Environment (getArgs)
-import System.Exit        (exitFailure)
-import Zinza              (Zinza (..), genericFromValueSFP, genericToTypeSFP, genericToValueSFP, parseAndCompileTemplateIO)
-
-import qualified Data.ByteString.Lazy.Char8 as LBS8
-import qualified Data.YAML                  as YAML
-
-main :: IO ()
-main = do
-    args <- getArgs
-    case args of
-        [src,tgt] -> do
-            run <- parseAndCompileTemplateIO src
-            -- this shouldn't fail (run-time errors are due bugs in zinza)
-            w <- run Z
-                { zJobs = -- version xenial flags                       old   needs     steps
-                          -------------------------------------------------------------------
-                    [ GhcJob "9.0.1"  False "--lib-only"                False ["8.8.4"] libSteps
-                    , GhcJob "8.10.4" False ""                          False ["8.8.4"] defSteps
-                    , GhcJob "8.8.4"  False "--solver-benchmarks"       False []        defSteps
-                    , GhcJob "8.6.5"  False "--complete-hackage-tests"  False ["8.8.4"] defSteps
-                    , GhcJob "8.4.4"  False ""                          False ["8.8.4"] defSteps
-                    , GhcJob "8.2.2"  False ""                          False ["8.8.4"] defSteps
-                    , GhcJob "8.0.2"  False "--lib-only"                False ["8.8.4"] libSteps
-                    , GhcJob "7.10.3" False "--lib-only"                False ["8.8.4"] libSteps
-                    , GhcJob "7.8.4"  False "--lib-only"                False ["8.8.4"] libSteps
-                    , GhcJob "7.6.3"  True  "--lib-only"                False ["8.8.4"] libSteps
-                    , GhcJob "8.8.4"  True  "--lib-only"                True  ["8.8.4"] $
-                        libSteps ++
-                        [ "lib-suite-extras --extra-hc /opt/ghc/7.0.4/bin/ghc-7.0.4"
-                        , "lib-suite-extras --extra-hc /opt/ghc/7.2.2/bin/ghc-7.2.2"
-                        , "lib-suite-extras --extra-hc /opt/ghc/7.4.2/bin/ghc-7.4.2"
-                        ]
-                    ]
-                , zMacosJobs =
-                    [ mkMacGhcJob "8.8.4" "https://downloads.haskell.org/~ghc/8.8.4/ghc-8.8.4-x86_64-apple-darwin.tar.xz"
-                    , mkMacGhcJob "8.6.5" "https://downloads.haskell.org/~ghc/8.6.5/ghc-8.6.5-x86_64-apple-darwin.tar.xz"
-                    ]
-                , zWinJobs =
-                    [ mkWinGhcJob "8.6.5"  Nothing           []
-                    , mkWinGhcJob "8.10.4" (Just "8.10.4")   []
-                    ]
-                , zMangleVersion = map mangleChar
-                , zOr            = (||)
-                , zNot           = not
-                , zLibOnly       = (== "--lib-only")
-                }
```

Comment:
Having every parameter (version, flags...) in a single place helped a lot when updating stuff... is there a way to do something similar natively?

Reply:
I tried different approaches. Typically one uses a [build matrix](https://docs.github.com/en/actions/using-jobs/using-a-build-matrix-for-your-jobs) but we seem to be cherry-picking desparate configurations without a clear meaning here.

A bit extreme but one could also build the matrix entierly manually by listing its entries (see [this](https://docs.github.com/en/actions/using-jobs/using-a-build-matrix-for-your-jobs#using-environment-variables-in-a-matrix) example, nevermind the use of environment variables). This would be close to what we do currently.

Also, I haven't yet opened `validate.sh` do see what it does, done that I can try to unpack and simplify this configuration.

Reply:
the matrix could be defined dynamically using external data (i would use the list of ghcs in the `tested-with` field of the .cabal file like `haskell-ci`), see https://www.cynkra.com/blog/2020-12-23-dynamic-gha/ for example